src/vcf/odb_report.cpp

```diff
@@ -0,0 +1,150 @@
+            flush();
+        } catch (std::exception &e) {
+            std::cerr << "An error occurred finalizing the error reporting: " << e.what() << std::endl;
+        }
+    }
+
+    void OdbReportRW::flush()
+    {
+        // possible recovery can be done here, ODB rollbacks automatically on error, and throws.
+        if (transaction.has_current()) {
+            transaction.commit();
+        }
+    }
+
+    // ReportWriter implementation
+    void OdbReportRW::write_error(Error &error)
+    {
+        error.get_severity() = Severity::ERROR;
+        write(error);
+    }
+    void OdbReportRW::write_warning(Error &error)
+    {
+        error.get_severity() = Severity::WARNING;
+        write(error);
+    }
+
+    void OdbReportRW::write(Error &error)
+    {
+        if (current_transaction_size == 0) {
+            // start transaction
+            transaction.reset(db->begin());
+        }
+
+        db->persist(error);
+
+        ++current_transaction_size;
+        if (current_transaction_size == transaction_size) {
+            // commit transaction
+            flush();
+            current_transaction_size = 0;
+        }
+    }
+
+    // ReportReader implementation
+    size_t OdbReportRW::count_warnings()
+    {
+        return count(odb::query<ErrorCount>::severity == Severity::WARNING);
+    }
+    void OdbReportRW::for_each_warning(std::function<void(std::shared_ptr<Error>)> user_function)
+    {
+        for_each(user_function, odb::query<Error>::severity == Severity::WARNING);
+    }
+
+    size_t OdbReportRW::count_errors()
+    {
+        return count(odb::query<ErrorCount>::severity == Severity::ERROR);
+    }
+    void OdbReportRW::for_each_error(std::function<void(std::shared_ptr<Error>)> user_function)
+    {
+        for_each(user_function, odb::query<Error>::severity == Severity::ERROR);
+    }
+
+    size_t OdbReportRW::count(odb::query<ErrorCount> query)
+    {
+        ErrorCount count;
+        if (transaction.has_current()) {
+            throw std::logic_error{"There's another transaction active. You can only read if the changes were flushed"};
+        } else {
+            transaction.reset(db->begin());
+//        size_t count = db->execute("SELECT COUNT(*) FROM Error");
+            count = db->query_value<ErrorCount>(query);
+            flush();
+        }
+        return count.count;
+    }
+    void OdbReportRW::for_each(std::function<void(std::shared_ptr<Error>)> user_function, odb::query<Error> query)
+    {
+        typedef odb::result<Error> result;
+
+        if (transaction.has_current()) {
+            throw std::logic_error{"There's another transaction active. You can only read if the changes were flushed"};
+        } else {
+            transaction.reset(db->begin());
+
+            result r{db->query<Error>(query + " ORDER BY " + odb::query<Error>::line)};
+
+            // TODO if result::iterator has operator*, operator!=, operator++, then use ranged for. not if we use shared_ptr
```

Comment:
doesn't apply anymore. First, the iterator was just accessed like `*i`, but now that we have to use `result::iterator::load()` I don't think we can use a ranged for. Comment removed in nex commit.